File: search-algorithms/sequential-search/jump-search/c++/jump-search.cpp
/*
worst case timing complexity = O (√n) 
*/

#include <iostream>
#include <math.h>

using namespace std;

int min(int left, int right) {
  return left > right ? right : left;
}

int jumpSearch(int vector[], int size, int wanted) {
  int step = sqrt(size);
  int prev = 0;

  while (vector[min(step, size)-1] < wanted) {
    prev = step;
    step += sqrt(size);

    if (prev >= size) return -1;
  }

  while (vector[prev] < wanted) {
    prev++;

    if (prev == min(step, size)) return -1;
  }

  if (vector[prev] == wanted) return prev;

  return -1;
}

int main() {
  int myVector[5] = { 10, 20, 40, 50, 30 };
  int result = jumpSearch(myVector, 5, 20);

  (result == -1) 
    ? printf("Elemento não encontrado no vetor\n")
    : printf("Elemento encontrado no índice: %d\n", result);

  return 0;
}

File: search-algorithms/sequential-search/jump-search/js/jump-search.js
function jumpSearch(array, size, wanted) {
  let step = Math.sqrt(size);
  let prev = 0;

  if (array[size-1] === wanted) return size-1;

  while (array[Math.min(step, size)-1] < wanted) {
    prev = step;
    step += Math.sqrt(size);

    if (prev >= size) return -1;
  }

  if (array[prev] < wanted) {
    prev++;

    if (prev === Math.min(step, size)) return -1;
  }

  if (array[prev] === wanted) return prev;

  return -1;
}

const array = [1, 2, 3, 4, 5, 6, 7, 8 ];
console.log(jumpSearch(array, 8, 7));